### PR TITLE
refactored code reading url from stdin

### PR DESCRIPTION
**Pull request by Jason23347**
This fixes issue #147
Mainly refactored code in text.c to read urls from stdin.
Using fgets() instead of scanf().
Changed search_t into linked list node.

### CODE REVIEW

File: src/axel.c [hunk 1]
```diff
@@ -66,9 +66,9 @@ static char *buffer = NULL;
 
 /* Create a new axel_t structure */
 axel_t *
-axel_new(conf_t *conf, int count, const void *url)
```
Comment (ismaell) [inline src/axel.c]: why do you need to remove the const qualifier?

**Comment by ismaell** [comment]
Overall looks good, but I would like that code to allocate the `search_t` first and read the URLs directly into `search_t.url`.

**Comment by Jason23347** [comment]
Thanks for code review, I will optimize the code  later.

File: src/search.c [hunk 2]
```diff
@@ -42,6 +42,7 @@
 
 #include <stdio.h>
 
+#include "config.h"
```
Comment (ismaell) [inline src/search.c]: This should go before `<stdio.h>`

**Comment by Jason23347** [comment]
Thank you very much for giving so many suggestions. It doesn't matter that asking me to make many changes. It's really nice of you to ensure code quality. I will try to make changes as soon as possible.

File: src/text.c [hunk 3]
```diff
@@ -234,94 +236,97 @@ main(int argc, char *argv[])
...
+			/* add URL list from stdin to end of list */
+			url_num += search_readlist(list + prealloc_num - 2, stdin) - 1;
+			continue;
 		}
+
+		/* add URL from argv */
+		strlcpy(tmp->url, argv[i], sizeof(tmp->url));
+		cur->next = tmp;
+		cur = tmp;
+		tmp++;
```
Comment (ismaell) [inline src/text.c:265]: This could be better expressed as:
```C
cur->next = tmp++;
cur = cur->next;
```

File: src/text.c [hunk 4]
```diff
@@ -234,94 +236,97 @@ main(int argc, char *argv[])
...
 		}
 
 		search_sortlist(search, i);
+
+		/* add 'search' into linked list */
+		j = min(j, conf->search_top);
+		for (i = 1; i < j; i++) {
+			search[i - 1].next = &search[i];
+		}
+		search[j - 1].next = list;
```
Comment (ismaell) [inline src/text.c:303]: Can you please move the sorting inside `search_sortlist`?

File: src/text.c [hunk 5]
```diff
@@ -234,94 +236,97 @@ main(int argc, char *argv[])
...
-		axel = axel_new(conf, argc - optind, search);
-		free(search);
-		if (!axel || axel->ready == -1) {
-			print_messages(axel);
-			goto close_axel;
-		}
+	tmp = search;
+	for (int i = 0; i < url_num; i++) {
+		printf("loaded: %s", tmp->url);
+		tmp = tmp->next;
```
Comment (ismaell) [inline src/text.c:321]: Debugging code?

Comment (ismaell) [reply src/text.c:321]: This should be guarded by `#ifdef DEBUG ... #endif`

File: src/text.c [hunk 6]
```diff
@@ -445,6 +450,20 @@ main(int argc, char *argv[])
 
  close_axel:
 	axel_close(axel);
+	/* free search here */
+	if (do_search)
+		free(search);
+ free_list:
+	cur = list[prealloc_num - 2].next;
```
[the remark this answers is not in the document]
Comment (Jason23347) [reply src/text.c:457]: Argv "-" and the last node were ignored so it should be like this...

[the remark this answers is not in the document]
Comment (ismaell) [reply src/text.c:457]: Right.

**Comment by ismaell** [comment]
I've saved this to the [url-from-stdin branch](https://github.com/axel-download-accelerator/axel/tree/url-from-stdin), and I made some small edits to it.

The reason I didn't merge it  is because there's some corner cases that don't work, e.g. when "-" is the only argument it doesn't fill properly, and I'm guessing there's a couple of issues if there's more of those arguments to be omitted.

**Comment by ismaell** [comment]
You may make a new PR with these issues solved, but if not, I'll try to solve them before v2.18 and merge it.

**Pull request closed without merging**